### DISCUSSION TITLE
KaTeX Matrix Brackets

### DISCUSSION THREAD

**Discussion by aaronjholbrook** [KaTeX/KaTeX#2742] [General]
I'm working on a project with a team on creating editable equations in SlateJS, and we've been trying to dig into how matrix brackets work in KaTeX. We've been struggling how it determines how they are sized, and if they could possibly be applied to something outside of KaTeX.

**Comment by ronkok** [comment]
My high-level description goes something like this.

For every glyph in the KaTeX fonts, KaTeX has detailed metrics for *height* (above the baseline) and *depth* (below the baseline). As KaTeX builds HTML, it keeps an accounting of the height and depth of each group that it constructs.

Using that height and depth data, KaTeX selects a delimiter from six options:

1. A glyph from the KaTeX_Main-Regular font      (small)
2. A glyph from the KaTeX_Size1-Regular font     (large)
3. A glyph from the KaTeX_Size2-Regular font     (large)
4. A glyph from the KaTeX_Size3-Regular font     (large)
5. A glyph from the KaTeX_Size4-Regular font     (large)
6. A delimiter constructed from pieces found in the KaTeX_Size4-Regular font (stack)
    
If you have forked KaTeX, you can see these glyphs by using https://opentype.js.org/glyph-inspector.html to view the `ttf` versions of these fonts in the `\KaTeX\submodules\katex-fonts\fonts` folder.

Delimiter size selection takes place in the `traverseSequence()` function of the `delimiter.js` module. The selection process is complicated by TeX size adjustments not only for font size but also for `textstyle | scriptstyle | scriptscriptstyle`. TeX does this in a confusing way, and since KaTeX emulates TeX, the KaTeX process is also less than straightforward. If you are writing editable equations, you might be able to just query the browser and get a size.

TeX creates stacked delimiters using glyphs that are vertically aligned very accurately. We have found that browsers sometimes are insufficiently accurate in this alignment and will sometimes render stacked delimiters with a one-pixel wide gap. So the next release of KaTeX will include PR #2698 using SVGs for the stretchy vertical parts of a stacked delimiter. 

There are probably people better qualified than me to describe all this. Two of the main KaTeX contributors have done some work on their own interactive math editors. That would be @kevinbarabash's [math-blocks](https://github.com/math-blocks/math-blocks) and @ylemkimon's [mathylem](https://github.com/ylemkimon/mathylem). You might take a look and see how they go about this.

Good luck!
